File: nodejs/continuousquery/src/continuous-query.js
//@ts-check
const {CacheManager,CacheItem,QueryCommand,ContinuousQuery,EventType,EventDataFilter,JsonDataType} = require('ncache-client');
const QueryDataModificationListener = require('ncache-client/src/client/QueryDataModificationListener');
const config = require('./app.config.json');
const {delay} = require("./helper/util");
let cache = null;
async function Run() { 
    // Initialize cache 
    await InitializeCache();

    //clear cache
    await cache.clear();
    // Create a simple products list
    const product = CreateNewProduct();
    // Adding item to cache
    await AddObjectToCache(product);
    // Build Query Command
    const queryCommand = BuildQueryCommand();

    // Register continuous query on cache
    const continuousQuery = await RegisterQuery(queryCommand);

    // Query keys in cache as per criteria
    await QueryKeysInCache(queryCommand);

    // Query data in cache as per criteria
    await QueryDataInCache(queryCommand);

    // Update an item in cache that is within query criteria to raise data modification event
    await UpdateObjectInCache(product);

    // Delete the existing object
    await DeleteObjectFromCache(product);

    //wait to recieve events
    await delay(1000);

    // Unregister query 
    await UnRegisterQuery(continuousQuery);
    
    //close cache
    await cache.close();
}
async function InitializeCache() {
    // Get Cache Name
    const cacheName = config.cacheProperties.cacheName
    if (!cacheName) {
        console.log("The CacheID cannot be null or empty.");
        return;
    }
    // Initialize an instance of the cache to begin performing operations:
    cache = await CacheManager.getCache(cacheName);

    // Print output on console
    console.log(`Cache ${cacheName} is initialized.`);
            
}
async function AddObjectToCache(product) {
    const cacheItem = new CacheItem(product,"Alachisoft.NCache.Sample.Data.Product");
    await cache.add(GetKey(product),cacheItem);
}
function BuildQueryCommand() {
    const query = "SELECT $Value$ FROM Alachisoft.NCache.Sample.Data.Product WHERE Category = ?";
    const queryCommand = new QueryCommand(query);
    let parametersMap = new Map();
    parametersMap.set("Category", "Edibles");
    queryCommand.setParameters(parametersMap);
    return queryCommand;
}
async function RegisterQuery(queryCommand) {
    const continuousQuery = new ContinuousQuery(queryCommand);
    // Add query notifications
    const listener = new QueryDataModificationListener(QueryDataModified);
    continuousQuery.addDataModificationListener(listener,[EventType.ItemRemoved], EventDataFilter.None);
    continuousQuery.addDataModificationListener(listener,[EventType.ItemUpdated], EventDataFilter.DataWithMetadata)

    // Register continuous query on server
    const messagingService = await cache.getMessagingService();
    await messagingService.registerCQ(continuousQuery);

    // Print output on console
    console.log("Continuous Query is registered.");

    return continuousQuery;
}
async function QueryKeysInCache(queryCommand) {
    const searchService = await cache.getSearchService();
    const cacheReader = await searchService.executeReader(queryCommand);
    if (cacheReader.getFieldCount() > 0) {
        while (await cacheReader.read()) {
            let key = cacheReader.getString(0);
            console.log("key: ",key);

            // A second call to fetch product details
            const product = await cache.get(key,JsonDataType.Object);
            PrintProductDetails(product);
        }
    }
}
async function QueryDataInCache(queryCommand) {
    const searchService = await cache.getSearchService();
    const cacheReader = await searchService.executeReader(queryCommand);
    if (cacheReader.getFieldCount() > 0) {
        while (await cacheReader.read()) {
            const product = cacheReader.getValue(1,JsonDataType.Object);
            PrintProductDetails(product);
        }
    }
}
async function UpdateObjectInCache(product) {
    // Modifiying and re-insert the item
    product.ClassName = "Class D";
    const cacheItem = new CacheItem(product,"Alachisoft.NCache.Sample.Data.Product");
    // This should invoke query RegisterUpdateNotification 
    await cache.insert(GetKey(product),  cacheItem);

    // Print output on console
    console.log("Item is updated in cache.");
}
async function DeleteObjectFromCache(product) {
    await cache.remove(GetKey(product), JsonDataType.Object);
    console.log('item is remove from cache');
}
async function  UnRegisterQuery(continuousQuery) {
    // UnRegister the continuous query on server
    const messagingService = await cache.getMessagingService();
    await messagingService.unRegisterCQ(continuousQuery);

    // Print output on console
    console.log("Continuous Query is unregistered.");
}
function QueryDataModified(_string, cQEventArg) {
     // Print output on console
     console.log("Continuous Query data modification event is received from cache.");
     switch (cQEventArg.getEventType())
     {
        case EventType.ItemAdded:
            console.log(`${_string} is added to cache`);
            break;
        case EventType.ItemRemoved:
            console.log(`${_string} is removed from cache`);
            break;
        case EventType.ItemUpdated:
            console.log(`${_string} is updated in cache`);
            break;
        default:
            break;
     }
}
function CreateNewProduct() {
    return({Id : 1, Name : "Dairy Milk Cheese", ClassName : "ClassA", Category : "Edibles",UnitPrice:12});
}
function PrintProductDetails(product) {
    if (product == null) return;
    console.log("\nId:           " + product.Id);
    console.log("Name:         " + product.Name);
    console.log("Class:        " + product.ClassName);
    console.log("Category:     " + product.Category);
    console.log("Unit price:   " + product.UnitPrice)
    console.log();

}
function GetKey(product) {
    return `Product:${product.Id}`
}
module.exports = {
    Run
}

File: python/continuous_query/continuous_query.py
import time
from ncache.client.CacheManager import CacheManager
from ncache.client.QueryCommand import QueryCommand
from ncache.client.ContinuousQuery import ContinuousQuery
from ncache.client.enum.EventType import EventType
from ncache.client.enum.EventDataFilter import EventDataFilter
from sample_data.product import Product


class ContinuousQuerySample:
    cache = None

    @staticmethod
    def run():
        # Initialize new cache
        ContinuousQuerySample.initialize_cache()

        # Generate instance of Product
        product = ContinuousQuerySample.create_new_product()
        key = ContinuousQuerySample.get_key(product)

        # Adding item to cache
        ContinuousQuerySample.add_object_to_cache(key, product)

        # Build Query Command
        query_command = ContinuousQuerySample.build_query_command()

        # Register continuous query on cache
        continuous_query = ContinuousQuerySample.register_query(query_command)

        # Query keys in cache as per criteria
        ContinuousQuerySample.query_keys_in_cache(query_command)

        # Query data in cache as per criteria
        ContinuousQuerySample.query_data_in_cache(query_command)

        # Update an item in cache that is within query criteria to raise data modification event
        ContinuousQuerySample.update_object_in_cache(key, product)

        # Delete the existing object
        ContinuousQuerySample.delete_object_from_cache(key)

        # Sleep for a while to allow events to be processed
        time.sleep(5)

        # Unregister query
        ContinuousQuerySample.unregister_query(continuous_query)

        # Dispose the cache once done
        ContinuousQuerySample.cache.close()

    @staticmethod
    def initialize_cache():
        cache_id = "demoCache"
        ContinuousQuerySample.cache = CacheManager.get_cache(cache_id)
        ContinuousQuerySample.cache.clear()
        print(f"\nCache '{cache_id}' is initialized.")

    @staticmethod
    def add_object_to_cache(key, product):
        ContinuousQuerySample.cache.add(key, product)
        print("\nItem is added to cache.")

    @staticmethod
    def build_query_command():
        query = "SELECT * FROM sample_data.product.Product WHERE id = ?"
        query_command = QueryCommand(query)
        parameters = {"id": "1001"}
        query_command.set_parameters(parameters)
        return query_command

    @staticmethod
    def register_query(query_command):
        continuous_query = ContinuousQuery(query_command)

        # Add query notifications
        continuous_query.add_data_modification_listener(ContinuousQuerySample.query_data_modified, [EventType.ITEM_ADDED], EventDataFilter.NONE)

        # Item update notification with DataWithMetadata as event data filter to receive modified item on updation
        continuous_query.add_data_modification_listener(ContinuousQuerySample.query_data_modified, [EventType.ITEM_UPDATED], EventDataFilter.DATA_WITH_META_DATA)

        # Item delete notification
        continuous_query.add_data_modification_listener(ContinuousQuerySample.query_data_modified, [EventType.ITEM_REMOVED], EventDataFilter.NONE)

        # Register continuous query on server
        ContinuousQuerySample.cache.get_messaging_service().register_cq(continuous_query)

        print("\nContinuous Query is registered.")
        return continuous_query

    @staticmethod
    def query_keys_in_cache(query_command):
        cache_reader = ContinuousQuerySample.cache.get_search_service().execute_reader(query_command, False, 512)
        if not cache_reader.get_is_closed():
            print("\nFollowing keys are fetched with Continuous Query.")
            while cache_reader.read():
                key = cache_reader.get_value(str, 0)
                print("Key: " + str(key))

    @staticmethod
    def query_data_in_cache(query_command):
        cache_reader = ContinuousQuerySample.cache.get_search_service().execute_reader(query_command, True, 512)
        if not cache_reader.get_is_closed():
            print("\nFollowing items are fetched with Continuous Query.")
            while cache_reader.read():
                id = cache_reader.get_value(str, None, "id")
                name = cache_reader.get_value(str, None, "name")
                unit_price = cache_reader.get_value(int, None, "unit_price")
                units_available = cache_reader.get_value(int, None, "units_available")
                print(f"\nID: {id}\nName: {name}\nUnit Price: {unit_price}\nUnits Available: {units_available}")

    @staticmethod
    def update_object_in_cache(key, product):
        product.units_available = 10
        ContinuousQuerySample.cache.insert(key, product)
        print("\nItem is updated in cache.")

    @staticmethod
    def unregister_query(continuous_query):
        ContinuousQuerySample.cache.get_messaging_service().un_register_cq(continuous_query)
        print("\nContinuous Query is unregistered.")

    @staticmethod
    def delete_object_from_cache(key):
        ContinuousQuerySample.cache.remove(key, Product)
        print("\nObject is deleted from cache.")

    @staticmethod
    def query_data_modified(key, cq_event_arg):
        print("\nContinuous Query data modification event is received from cache.")
        event_type = cq_event_arg.get_event_type()
        if event_type == EventType.ITEM_ADDED:
            print(key + " is added to cache.")
        elif event_type == EventType.ITEM_REMOVED:
            print(key + " is removed from cache.")
        elif event_type == EventType.ITEM_UPDATED:
            print(key + " is updated in cache.")
            if cq_event_arg.get_item() is not None:
                product = cq_event_arg.get_item().get_value(Product)
                print(product)

    @staticmethod
    def create_new_product():
        product = Product(id="1001", name="Tea", unit_price=50, units_available=20)
        return product

    @staticmethod
    def get_key(product):
        key = f"Product:{product.id}"
        return key


ContinuousQuerySample.run()
